src/java_bytecode/generate_java_generic_type.cpp

```diff
@@ -189,40 +207,129 @@ irep_idt generate_java_generic_typet::build_generic_tag(
   const java_class_typet &original_class) const
 {
   std::ostringstream new_tag_buffer;
-  new_tag_buffer << original_class.get_tag();
-  new_tag_buffer << "<";
-  bool first=true;
-  for(const typet &type_argument : existing_generic_type
-    .generic_type_arguments())
+  const std::string &original_class_tag = original_class.get_tag().c_str();
+  auto generic_argument_p =
+    existing_generic_type.generic_type_arguments().begin();
+
+  // if the original class is implicitly generic, add type
+  // arguments for all generic outer classes
+  if(is_java_implicitly_generic_class_type(original_class))
   {
-    if(!first)
-      new_tag_buffer << ", ";
-    first=false;
-
-    INVARIANT(
-      !is_java_generic_parameter(type_argument),
-      "Only create full concretized generic types");
-    const irep_idt &id(id2string(type_argument.subtype().get(ID_identifier)));
-    new_tag_buffer << pretty_print_java_type(id2string(id));
-    if(is_java_array_tag(id))
+    const java_implicitly_generic_class_typet
+      &implicitly_generic_original_class =
+        to_java_implicitly_generic_class_type(original_class);
+    const auto &implicit_generic_types =
+      implicitly_generic_original_class.implicit_generic_types();
+
+    std::string current_outer_class_name =
+      original_class_tag.substr(0, original_class_tag.find("$"));
+    new_tag_buffer << current_outer_class_name;
+    bool first_outer = true;
+    bool first = true;
+
+    for(const auto &parameter : implicit_generic_types)
     {
-      const typet &element_type =
-        java_array_element_type(to_symbol_type(type_argument.subtype()));
-
-      // If this is an array of references then we will specialize its
-      // identifier using the type of the objects in the array. Else, there can
-      // be a problem with the same symbols for different instantiations using
-      // arrays with different types.
-      if(element_type.id() == ID_pointer)
+      PRECONDITION(is_java_generic_parameter(parameter));
+      const std::string &parameter_name = to_java_generic_parameter(parameter)
+                                            .type_variable()
+                                            .get_identifier()
+                                            .c_str();
+      // get the class name without the `java::` prefix, e.g., for
+      // `java::Class::T` parameter_class_name is `Class`
+      const std::string &parameter_class_name =
```

Comment:
I found it impossible to believe this isn't already a utility, but the closest i found is `pretty_print_java_type` in `java_utils`, not sure this is exactly right though.  It might be preferable, rather than directly manipulating the string is to store the class name inside the `java_generic_parametert`? 

If not - I suggest removing some magic numbers (e.g. `const size_t prefix_length="java::".length()`) and adding some conditions like `has_prefix("java::")`.

Reply:
Will this remove packages? I think we don't want to do that in general since users may have two classes whose name only differ by packages. This would be good to add a test for though possibly out of scope for this task.

Reply:
I added a function to `java_generic_parametert` to extract the class name. And I also made sure that the package name is not removed - it was a one-line-change.